Add Tide prompt

IlanCosman opened the pull request:
Hi, this PR adds [Tide](https://github.com/IlanCosman/tide) to the prompts sections.

IlanCosman commented:
Gentle bump to @jorgebucaran 😄. Any updates on this?

IlanCosman commented:
Hi @jorgebucaran 👋 , would you be open to adding another maintainer to this repository? I know these pull requests aren't very important since there are no security flaws or critical bugs in your code 😆, but I think the fish community would benefit from developers receiving a timely response to their work.

I understand the importance of vetting prospective additions--you're attaching your name to them in a way--but a month or more on some of these PRs with no responses or after everything has been cleared seems a bit much.

Awesome-fish is rightfully one of the first things that pops up on google when you search "best fish themes" or "best fish extensions" etc, and as such serves to introduce new users to the best that fish has to offer. Thus, this list has a vested interest in including all the coolest projects and being up to date.

Since this repo doesn't get too many PRs, perhaps a veteran community member like faho, ridiculousfish etc, would be willing to help. I'm also happy to to assist. 😄

I greatly admire the work you've put into the fish community and I hope you'll take this suggestion into consideration.

jorgebucaran commented:
@IlanCosman Thank your for your feedback and let me check the current PR situation. The reason some PRs are still open is because I felt they just weren't "awesome" enough. I know that's terribly opinionated, but what's the point of this project? To be a curation, not just a bunch of fish projects in a markdown list. Leaving those PRs open without any info was bad and I should've closed them right away or provided more info / followed them up, but I should be done with that by the time you get to this message.

Tide has potential, and I know someone has been posting periodical updates in r/fishshell to get community traction. These were my initial thoughts:

- You could improve the presentation of the project. While you did a good job with the screenshots and GIFs, they're low resolution and so many of them tend to be overwhelming as well.
- What do you mean by Pure compatibility? Pure is a different fish prompt.
- I'd like to see the LICENSE information in the README as well.
- I'd like to see info about how to manage Tide using Oh My Fish and Fisher, since there's a number of people who use them to manage their shell already. 
- I don't understand Tide's file hierarchy. There seem to be fish files in all the unexpected places! 😄 
  - You are using conf.d snippets to [preload fish functions](https://github.com/IlanCosman/tide/tree/master/conf.d), but fish already has a function autoload mechanism that you should rely upon instead.
  - My personal opinion: I feel Tide is more complicated than it needs to be.

> Since this repo doesn't get too many PRs, perhaps a veteran community member like faho, ridiculousfish etc, would be willing to help. I'm also happy to to assist. 😄

I surely appreciate all the help I can get. Other than merging rights, you can already review PRs and leave your comments. Why don't you start there? If you really want to help, here's one thing you could do right now: help me check if there are any projects that should be taken down—maybe they're not awesome anymore, were rewritten in a different language (I only want to curate projects written in fish) or shouldn't have been there to begin with.

IlanCosman commented:
Thanks for the speedy response @jorgebucaran, and thanks for dealing with those stale PRs. 😄 That was me posting in the subreddit.

Addressing your points in order:

- I'm not sure what you mean by the images being low resolution, I can zoom in to 150% and they are a bit fuzzy. Perhaps you could elaborate on that? The videos are animated svgs and therefore infinite resolution.

  - That's a good point about the number of animations/images in one area, I'll see if I can improve that!

- I thought the Pure Compatibility feature was fairly well explained:

  - > Tide can produce the same prompt as Pure.
  - This means that users of Pure should consider switching to tide for the benefits it provides (multi-line right prompt, wizard, one line left prompt, stash indicator, speed etc.)
    - Perhaps this was not sufficiently explained in the README

- I think that the the badge in the title displaying the license as well as github displaying it in the righthand section is plenty in terms of licensing, but if you strongly feel that it should be added to the bottom of the README then I will do so.

- I haven't researched making Tide work with Oh My Fish, I should do so. I will discuss fisher support at the end.

- An explanation of Tide's strange file structure:

  - The tide_theme folder contains the `tide` command's subcommands. More importantly it contains a number of static configuration files that are only accessed through the `tide configure` command. It also contains a copy of Tide's `fish_prompt`. If the user changes the prompt and runs the wizard all the way through, the user's fish_prompt will be overwritten with Tide's (there is a warning).
  - I use the conf.d folder for functions that will be needed immediately on startup. While I could put them in the functions folder, there would be no impact on startup time as they are run immediately. As for the two simple snippets relating to cursor movement, I just put them in one file to have less file pollution, a bit silly.
  - It basically comes down to this: If I were to move every file I can into the functions folder, there would be almost 50 Tide files in there. Currently there are 13. That would make it harder for the user to use their functions folder for personal stuff and for the developer to categorize things.
  - That being said, I'm not about to die on this hill. Im completely open to feedback on this matter from those more experienced than myself. The fact is though that Tide has a lot of files and that number is only going to be increasing.
  - Given this, I would prefer to use a system with directories and structure rather than tossing everything into the functions folder.

- Fisher support:
  - Given the current file structure of Tide, I'm not sure how to support fisher. I don't know enough about fisher, but I don't think it can do just random directories like tide_theme or saving a fish_prompt file elsewhere for safe keeping (though I could just rename it yes).
  - Having my own install function also allows me to conditionally install things for development like tests, though this is not particularly important.

I would absolutely appreciate your thoughts on these subjects Jorge.

This was silly of me but I didn't realise others were encouraged to weigh in on what counts as awesome or not, I'll certainly do so in the future. I'll definitely be making some awesome-fish PRs soon as well.

jorgebucaran commented:
> I'm not sure what you mean by the images being low resolution, I can zoom in to 150% and they are a bit fuzzy. Perhaps you could elaborate on that? The videos are animated svgs and therefore infinite resolution.

Ah, you're right. I felt they could be more interesting, it kind of looks like fish default theme, so I went with poor screenshots/GIFs. The resolution is fine. SVG videos are a really nice trick I learned about not long ago. 👍

IlanCosman commented:
@jorgebucaran can we continue this conversation? I'd like to improve my project and merge here 😄

jorgebucaran commented:
> @IlanCosman Given the current file structure of Tide, I'm not sure how to support fisher. I don't know enough about fisher, but I don't think it can do just random directories like tide_theme or saving a fish_prompt file elsewhere for safe keeping (though I could just rename it yes).

Have you considered "not doing directories"? What about putting all your functions under `functions`, completions under `completions` and snippets under `conf.d`?

IlanCosman commented:
I addressed that a few lines above. (paraphrased somewhat)

> If I were to move every file I can into the functions folder, there would be almost 50 Tide files in there. Currently there are 13. That would make it harder for the user to use their functions folder for personal stuff and for the developer to categorize things.
> Tide has a lot of files and that number is only going to be increasing. Given this, I would prefer to use a system with directories and structure rather than tossing everything into the functions folder.

To be honest it seems ridiculous to me to put everything in the functions folder when the _majority_ of my functions have no meaning outside of certain Tide controlled contexts. That would simply hamper organization for the sake of supporting fisher. I _do_ want to support fisher but it feels like fisher was simply not made for a project as complex as the one I'm trying to create.

jorgebucaran commented:
> feels like fisher was simply not made for a project as complex as the one I'm trying to create.

This is probably true, and for a reason—fisher doesn't try to be more complex than the fish shell itself and fish is fairly simple. Your file organization is useful just for that, organizing files. Every function will run in the same scope because fish lacks private function scope / namespaces. Maybe you're using the wrong language.

Having said that, I believe your project deserves to be in the awesome list, so let's just address the presentation/documentation. 

> Tide is a Fish theme that provides a powerful, flexible, and out-of-the-box experience.

Could we work on the title a bit? 

- Tide seems more like a prompt manager since you can have different prompts (or themes if you prefer to call them that).
- What does this mean? "powerful, flexible, and out-of-the-box experience". It doesn't really us much about Tide.

IlanCosman commented:
Let me get back to you on this after I release version 2.0 in a few days or so. That's going to involve a fairly radical shift in the documentation anyways. Good point though on the "prompt manager" thing, I'll see how I can fit that into the documentation.

jorgebucaran commented:
@IlanCosman Awesome, thank you!

jorgebucaran commented:
@IlanCosman Friendly pin just to see how are things over there! 🌊

IlanCosman commented:
> version 2.0 in a few days or so

😂 

It's actually going really well, just took a bit longer than expected. All of the code has been written for powerline so Today is going to be documentation day. I have a gorgeous new logo and the Readme is hopefully going to look sick.
